### DISCUSSION TITLE
User input specification

### DISCUSSION THREAD

**Discussion by GregoryAlbouy** [benchttp/runner#13] [Features decisions]
Let's determine specification for the user input / config, so we can implement the parsing with proper validation.

**Comment by c1-ra** [comment]
For reference, what we had imagined when brainstorming first:
![run config](https://user-images.githubusercontent.com/74239400/150685241-5abb5151-da0c-4d30-a4be-889349f7c06b.png)
and
![runner-io2](https://user-images.githubusercontent.com/74239400/150686266-1b982959-6a20-4828-aa0b-061ef7177508.png)


I think we could do something like this:


### Test config / Parameters:
**requestsNumber** (I find it clearer than Count): int
	Number of requests to run, use duration as exit condition if omitted
	Optional. -1 or > 0.
	We could add a maximum but I do not think it is necessary, we can let the user choose
	Flag -r

**interval**: int
	Interval between each request in ms
	Optional. -1 or > 0. It should not be > duration.
	(Flag -i?)

**totalTimeout**: int
	Duration of test, in seconds
	Mandatory. Default 60. Must be > 0.
	Flag -d

**timeoutPerRequest**: int
	Timeout for each http request, in seconds
	Mandatory. Default 10. Must be > 0.
	Flag -t

**concurrentRequestsNumber**: int
	Do the user wants to use concurrent requests, and how many ones
	Mandatory. Default: 1. Must be > 0. 1 means only one thread so actually no concurrent requests.
	Flag -c

(
Eventually, can be added later:
**excludeReadingResponseBodyFromTimeout**: bool
	Exclude reading of the response body in the request timeout
	Mandatory. Default: False.
	Flag -e would mean we want to set it to True
)

### curl config / Request:
**url**: string
	Url of the endpoint to be tested
	Mandatory. Should be a valid Url.

**method**: string
	Method to test
	Mandatory. Default: GET. Should be a valid HTTP request method.

**headers**: string (JSON)
	JSON with the headers of the request to be tested.
	Optional. Should be in JSON and each data a valid HTTP header and a valid value for it.
	Maybe we should first limit the number of headers we offer to test with the most common ones.

**data**: string (JSON)
	JSON representing the data to be sent
	Optional. Only valid for specific methods.

### Process config / metadata:
**label**: string
	Name the user wants to use to label the test
	Mandatory. Default: compute something from the config already provided? 
	There should be no duplicates (how to manage this? Should we just not allow duplicates and ask the user for another one, or automatically add an incremental number to the label and run the test anyway?)

**tags**: string (JSON)
	Array of strings in JSON. Tags the user want to add to the test.
	Optional.
